src/com/phonebookservice/database/ContactsDatabase.java

```diff
@@ -0,0 +1,69 @@
+package com.phonebookservice.database;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+
+import com.phonebookservice.model.Contact;
+import com.phonebookservice.util.MyArrayList;
+
+public final class ContactsDatabase {
+    private static ContactsDatabase singleton;
+    private final ContactsFileParser parser;
+    private final Collection<Contact> contactList;
+    private final Map<String, Contact> idToContactMap;
+
+    /**
+     * Contacts database constructor.
+     *
+     * @param fileName the file name.
+     */
+    private ContactsDatabase(final String fileName) {
+        this.contactList = new MyArrayList<Contact>();
+        this.idToContactMap = new HashMap<>();
+        this.parser = new ContactsFileParser(fileName);
+        this.readContacts();
+    }
+
+    /**
+     * Get instance of file data access adapter.
+     *
+     * @param fileName the file name.
+     * @return contacts database.
+     */
+    public static ContactsDatabase getInstance(final String fileName) {
+        if (singleton == null) {
+            singleton = new ContactsDatabase(fileName);
+        }
+
+        return singleton;
+    }
+
+    private void readContacts() {
+        final Collection<Contact> contacts = this.parser.readContacts();
+        this.contactList.addAll(contacts);
+
+        for (Contact contact : contacts) {
+            this.idToContactMap.put(contact.getId(), contact);
+        }
+    }
+
+    /**
+     * get contact.
+     *
+     * @param contactId the contact id.
+     * @return contact.
+     */
+    public Contact get(final Long contactId) {
+        return this.idToContactMap.get(contactId.toString());
```

Comment:
What happen if Id is null??
What happen if id doesn't exist ??

Reply:
if not exist in map will return null no problem
if null, there is already a check in filedataAccessAdapter. but better to do it here in this method

Reply:
You are using Map interface.
HashMap accept null, so when u do get(null) will return null. However Hashtable doesn't accept null and when u do get(null) it will throw null pointer exception